src/main/java/net/snowflake/client/jdbc/SnowflakeDatabaseMetaData.java

```diff
@@ -1127,8 +1131,18 @@ public ResultSet getProcedureColumns(
     while (resultSetStepOne.next()) {
       String procedureNameUnparsed = resultSetStepOne.getString("arguments").trim();
       String procedureNameNoArgs = resultSetStepOne.getString("name");
+      String schemaName = resultSetStepOne.getString("schema_name");
+      // Check that procedure name and schema name match the original input in case wildcards have
+      // been used.
+      // Procedure name column check must occur later when columns are parsed.
+      if ((compiledProcedurePattern != null
+              && !compiledProcedurePattern.matcher(procedureNameNoArgs).matches())
+          || (compiledSchemaPattern != null
+              && !compiledSchemaPattern.matcher(schemaName).matches())) {
+        continue;
+      }
       String showProcedureColCommand =
-          getSecondResultSetCommand(catalog, schemaPattern, procedureNameUnparsed, "procedure");
```

Comment:
Current schema name should be used to get 2nd set of results rather than original schema pattern, which is too broad and leads to redundant, extra rows.

Reply:
Do you mind giving some explanations on the duty of both the first step and the second step?

Reply:
Sure:
1. A show command is issued, like `show /* JDBC:DatabaseMetaData.getProcedures() */ procedures in database "TESTDB"`. If there are wildcards in the schema, we will search for procedures directly in the database instead of narrowing down by schema. `resultsetStepOne` is returned.
2. With the new logic I added in on lines 1135 to 1143, the schema and procedure names are fetched from each row of `resultsetStepOne` and if the schema and procedure name do not match the input arguments, that row is skipped
3. For rows with matching schemas and procedure names, a describe command is issued for the procedure name, like `desc procedure TESTDB."SCH1"."STPROC1"(FLOAT, VARCHAR);`. `resultSetStepTwo` is returned.
4. The results of the describe command are parsed and if the procedure name column does not match the input argument, the row is skipped. If the column name does match, a row is created from the info from both resultsetStepOne and resultsetStepTwo and returned as part of the final resultset that this function returns.

So, the bug here was two-fold:
1. The schema pattern from the function input argument was directly used in the describe command, even if it contained wildcards. This didn't work because if someone created a schema TEST1 and then queried getProcedureColumns() with a schema name of TEST_, the underscore should have matched the "1" and returned results for that schema. The wildcards don't work for desc commands so the match didn't occur. Now the actual schema name is used.
2. There was no logic to narrow down the results of resultsetStepOne in the event a wildcard schema was used. We searched for all procedures in the database, across all schemas (due to the schema wildcard) and failed to narrow down the results by schema afterwards. This wasn't noticeable due to problem #1 above.

Reply:
Sounds great, thanks for the detailed explanation!